### PR TITLE
python311Packages.cython: restrict patch to python versions where the API exists

### PR DESCRIPTION
**Pull request by voidus**
This doesn't actually cause `python311Packages.Cython` to fail it's build but `python311Packages.skia-pathops`

###### Description of changes

Cython used internal python APIs for a while, but that was removed at some point. We added a patch to add that back in in e69d0a0aa3bf but since the API was removed in python3.11 we cannot use the patch anymore.

I can't confidently say I understand why the patch was added so I don't want to speak on whether it is still needed, but since it doesn't work on python3.11 we should probably merge this?

Pinging
- @collares Since you added the patch
- @FRidh as Cython maintainer
- @barinov274 as skia-pathops maintainer

###### Things done

Only use the patch on python version <=3.11

- Built on platform(s)
  - [x] x86_64-linux (Cython and skia-pathops)
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by collares** [comment]
Can you apply https://github.com/cython/cython/commit/e337825cdcf5e94d38ba06a0cb0188e99ce0cc92 atop the present patch instead of removing it? It effectively reverts the changes in `Cython/Utility/ExtensionTypes.c` (because they were upstreamed) but does not revert the `Cython/Compiler/*` changes, meaning that Cython will now still use trashcan support (but the upstream version instead).

**Comment by collares** [comment]
For reference, the original patch (the one that would be removed if this PR were merged right now) was committed as https://github.com/cython/cython/commit/f781880b6780117660b2026caadf4a6d7905722f. I'm pointing this out because the patch is currently fetched from Sage's repository, which might make code archaeology unnecessarily hard.

**Comment by collares** [comment]
You'll probably want to target `staging` (or `python-updates`, if there's an open Python Updates PR, but I couldn't find one at a quick glance), because Cython is a transitive dependency of many many packages.

**Comment by collares** [comment]
> I thought it was uppercase `Cython` for some reason, maybe we need to trigger a rebuild? I'm not sure how ofborg works.

ofBorg looks at the commit title, not at the PR title. It will automatically trigger a new build on push (even force pushes).

**Comment by voidus** [comment]
Added the second patch as you suggested. I'm also pulling in both of them from the official repo, let me know/revert the lines if you'd prefer to leave the old one as it is.

@ofborg build python311Packages.skia-pathops

File: pkgs/development/python-modules/Cython/default.nix
Comment (collares) [inline pkgs/development/python-modules/Cython/default.nix:54]:
```suggestion
    # The above commit introduces custom trashcan macros, as well as
    # compiler changes to use them in Cython-emitted code. The latter
    # change is still useful, but the former has been upstreamed as of
    # Python 3.8, and the patch below makes Cython use the upstream
    # trashcan macros whenever available. This is needed for Python
    # 3.11 support, because the API used in Cython's implementation
    # changed: https://github.com/cython/cython/pull/4475
```

**Review by collares — APPROVED** [review]
I've extended the comment a bit, but I am happy with this change if it fixes the skia-pathops problem. Please switch the target branch to `staging`, though.

**Comment by voidus** [comment]
Done and done :)

**Comment by collares** [comment]
Great, many thanks for updating the patches! And thanks for pinging me in the first place, I probably would have missed it. Let's wait for the Cython maintainer's approval, but please ping me if that hasn't happened in a reasonable timeframe (a week, say).

**Comment by SuperSandro2000** [comment]
fails to build:

```
gcc -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -I/nix/store/jwg037i9jji90ap1112wd5hb5f4lg9iy-libxcrypt-4.4.28/include -fPIC -I./cypari2 -I/nix/store/rhqyzfxyk43gmr4yz1w9b7b8h2mz4r4m-python3.11-cysignals-1.11.2/lib/python3.11/site-packages/cysignals -I/nix/store/zy158n41isy1l8asc4q6kgz4bj255py8-pari-2.13.4/include -I/nix/store/s2q0x350276vgsqls8ws317kcrdii2xk-python3-3.11.0/include/python3.11 -c cypari2/convert.c -o build/temp.linux-x86_64-cpython-311/cypari2/convert.o
cypari2/convert.c: In function ‘__pyx_f_7cypari2_7convert_PyLong_FromINT’:
cypari2/convert.c:3395:21: error: lvalue required as unary ‘&’ operand
 3395 |   __pyx_v_sizeptr = &Py_SIZE(((PyObject *)__pyx_v_x));
      |                     ^
error: command '/nix/store/mjfccdxzpb87wa4pz56fkv6l63pw9in4-gcc-wrapper-11.3.0/bin/gcc' failed with exit code 1
error: builder for '/nix/store/g5ibxzw5d297q3n36qrvgv2n5zw1lsff-python3.11-cypari2-2.1.2.drv' failed with exit code 1;
```

**Comment by collares** [comment]
@SuperSandro2000 Thanks for the heads up! I tested cypari2 out of an abundance of caution, but it turns out to be incompatible with Python 3.11 for reasons unrelated to this PR (https://github.com/sagemath/cypari2/issues/114). It was fixed upstream (https://github.com/sagemath/cypari2/pull/120), but I think we can afford to wait for a new cypari2 release since Python 3.11 is opt-in in Nixpkgs for now. Either way, the compilation problem is caused solely by Python 3.11 and not by this PR.

**Comment by SuperSandro2000** [comment]
@ofborg build python311Packages.cython

**Pull request merged**
